### ISSUE TITLE
A generic method cannot define an instance of its containing interface

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-compiler#2072]
The problem occurs when a method attempts to define an implementation of its interface, and the method has a parameter that involves a type parameter of the interface.

This:

```ceylon
interface Foo<T> {
    shared Foo<U> map<U>(T f)
        =>  object satisfies Foo<U> {};
}
```

fails with:

```
anonymous_0_ is not abstract and does not override abstract method <U>map(com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor,U) in Foo - Ceylon Backend Error
```

This *does* work:

```ceylon
interface Foo<T> given T satisfies Object {
    shared Foo<U> map<U>(Callable<U, [T]> f)
            given U satisfies Object 
        =>  object satisfies Foo<U> {};
}
```

which is fairly common, but this slight variation doesn't:

```ceylon
interface Callable2<out Return, in Arguments> 
        given Arguments satisfies Anything[] {}

interface Foo<T> given T satisfies Object {
    shared Foo<U> map<U>(Callable2<U, [T]> f)
            given U satisfies Object 
        =>  object satisfies Foo<U> {};
}
```

**Comment by tombentley** [comment]
The issue here is that when we generate the implementation of `map()` in the companion class we use the type parameter name `U` but then redeclare a `U` in the implementation of the anonymous class's `map()` but in the same declaration we're still using `U` to refer to the outer classes `U`. 

We can see the same thing going on if `Foo` has the following member class:

    class C<U>() satisfies Bug2072<U> {}

(Note there that `C`'s type parameter, `U`, has to be the same as `map()`'s).

I suppose the solution is to rename the masking `U` to something else, and use the same name consistently.

**Comment by FroMage** [comment]
Do you have sample Java code to illustrate?

**Comment by tombentley** [comment]
Well cutting out a lot of the boiilerplate, the gist of it looks like this:

```

final class Bug2072$impl<T> {
    public <U>.com.redhat.ceylon.compiler.java.test.issues.bug20xx.Bug2072<U> map(final .com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor $reified$U, final T f) {
        return (.com.redhat.ceylon.compiler.java.test.issues.bug20xx.Bug2072<U>)(.com.redhat.ceylon.compiler.java.test.issues.bug20xx.Bug2072)(
            let
            {
                
                class anonymous_0_ implements .com.redhat.ceylon.compiler.java.runtime.model.ReifiedType, .com.redhat.ceylon.compiler.java.test.issues.bug20xx.Bug2072<U> {

                    @.java.lang.Override
                    public final <U>.com.redhat.ceylon.compiler.java.test.issues.bug20xx.Bug2072<U> map(final .com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor $reified$U, final U f) {
                        return $com$redhat$ceylon$compiler$java$test$issues$bug20xx$Bug2072$this$.map($reified$U, f);
                    }

            returning new anonymous_0_();
        );
    }
}
```

See how the inner `map()` declares `U` masking the one declared by the outer `map()`? Well the `final U f`  in the inner `map()` is actually the outer `map()`s `U`, so java concludes the method hasn't been implemented.

**Comment by tombentley** [comment]
This is actually going to be difficult to fix because as far as the Ceylon model is concerned there is only one `TypeParameter` so it's not like I can construct a non-ambiguous name for the type parameter -- the disambiguation is needed depending on context within the code generator, and that context flows down to into `makeJavaType()` because there could be `ProducedType`s which need to be adjusted.

**Comment by FroMage** [comment]
I don't follow. You're saying the inner `map.f` is resolved to the outer `map`'s `U`??

**Comment by tombentley** [comment]
No, the other way around. It must be resolved to the outer `map`'s `U` for javac to see that `map` has been implemented by the `anonymous_0_` class, but due to the masking caused by the inner `map`s `U` it resolves to that inner `U` and javac decides the method has not been implemented.

**Comment by FroMage** [comment]
OK so this one is a lot tougher than it seems, because:

```ceylonn
interface Foo<T> {
    shared Foo<U> map<U>(T t, U u)
        =>  object satisfies Foo<U> {};
}
```

Which gives us a `ProducedTypedReference` for `Foo.map.#1.map` of type `<U> Foo<U>(U t, U u)` which we need to rename to `<RenamedU> Foo<RenamedU>(U t, RenamedU u)` because only the first parameter is substituted, so we need to keep track of which parts of a `ProducedType` has come from type argument substitution. I have a fix for that in a branch but fixing this showed errors in `nonWideningType` in `empty` which returned to me a type for `Empty.map` that was `Iterable<Iterable.map.Result>` rather than `Iterable<Empty.map.Result>` (it is using the refined decl type parameters) which only works by accident I guess, since the type parameters of the mixin bridge methods take their name from the refined decl, so it works, but I can't see that the type parameters have been renamed.

**Comment by tombentley** [comment]
>  so we need to keep track of which parts of a `ProducedType` has come from type argument substitution.

This might be useful for fixing #2051 too.

**Comment by FroMage** [comment]
Let's move it to 1.3, it's really hard to fix.
